### PR TITLE
JCLOUDS-1231: Implement the SecurityGroupExtension in ARM

### PR DESCRIPTION
**Pull request by nacx**
https://issues.apache.org/jira/browse/JCLOUDS-1231

This implements the SecurityGroupExtension and adds support for the `TemplateOptions.inboundPorts` feature. There is also some minor refactoring in the cleanup classes, to make it more reusable and to minimize the number of times the code is executed when cleaning up orphaned nodes.

All extension live tests are passing.

/cc @andreaturli

### CODE REVIEW

**Comment by nacx** [comment]
@andreaturli I've just added all the mentioned code cleanup and introduced the resource group cache. Please have a look.

All extension live tests are green. I'll run the compute service adapter ones in a while.

```bash
Starting test testCreateSecurityGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-1] Test testCreateSecurityGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 1215ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
Starting test testCreateNodeWithInboundPorts(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-2] Test testCreateNodeWithInboundPorts(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 325603ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
Starting test testCreateNodeWithSecurityGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-2] Test testCreateNodeWithSecurityGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 475649ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
Starting test testGetSecurityGroupById(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-2] Test testGetSecurityGroupById(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 27745ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
Starting test testAddIpPermission(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-3] Test testAddIpPermission(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 3496ms
Test suite progress: tests succeeded: 5, failed: 0, skipped: 0.
Starting test testRemoveIpPermission(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-4] Test testRemoveIpPermission(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 3441ms
Test suite progress: tests succeeded: 6, failed: 0, skipped: 0.
Starting test testAddIpPermissionsFromSpec(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-5] Test testAddIpPermissionsFromSpec(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 1750ms
Test suite progress: tests succeeded: 7, failed: 0, skipped: 0.
Starting test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-5] Test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 7, failed: 0, skipped: 1.
[pool-1-thread-5] Test testRemoveIpPermissionWithCidrExclusionGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 7, failed: 0, skipped: 2.
Starting test testDeleteSecurityGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest)
[pool-1-thread-5] Test testDeleteSecurityGroup(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest) succeeded: 705ms
Test suite progress: tests succeeded: 8, failed: 0, skipped: 2.
Tests run: 10, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 970.066 sec - in org.jclouds.azurecompute.arm.compute.extensions.AzureComputeSecurityGroupExtensionLiveTest

Results :

Tests run: 10, Failures: 0, Errors: 0, Skipped: 2
```

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeSecurityGroupExtension.java [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+      if (networkSecurityGroup == null) {
+         throw new IllegalArgumentException("Security group " + group.getName() + " was not found");
+      }
+
+      NetworkSecurityRuleApi ruleApi = api.getNetworkSecurityRuleApi(resourceGroup.name(), networkSecurityGroup.name());
+      int nextPriority = getRuleStartingPriority(networkSecurityGroup);
+
+      for (String ipRange : ipRanges) {
+         NetworkSecurityRuleProperties properties = NetworkSecurityRuleProperties.builder()
+               .protocol(Protocol.fromValue(protocol.name())) //
```
Comment (andreaturli) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeSecurityGroupExtension.java]: please remove `//`

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeSecurityGroupExtension.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+      this.locations = locations;
+      this.securityGroupConverter = groupConverter;
+      this.securityGroupAvailable = securityRuleAvailable;
+      this.resourceDeleted = resourceDeleted;
+      this.resourceGroupMap = resourceGroupMap;
+   }
+
+   @Override
+   public Set<SecurityGroup> listSecurityGroups() {
+      return ImmutableSet.copyOf(concat(transform(locations.get(), new Function<Location, Set<SecurityGroup>>() {
```
Comment (nacx) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeSecurityGroupExtension.java]: I think it is good to enforce immutability. All the ARM model is immutable. Why should these collections be different? There are no methods that receive a list of security groups, so I don't see any value a mutable list could have, so let's try to default to immutable as a best practice.

**Comment by andreaturli** [comment]
+1 @nacx, please merge it!

**Comment by nacx** [comment]
Pushed to master and 2.0.x!

**Pull request closed without merging**